# The Weather Set. 

The weather data set is a time-series data set with per-hour information about the weather conditions at a particular location. It records the following: Temperature, Dew Point Temperature, Relative Humidity, Wind Speed, Visibility, Pressure, and Conditions. 

This data is available as a CSV File and I am doing this project via the Data Science Lovers channel on YouTube (https://youtu.be/4hYOkHijtNw?si=mo01ack8LxgB2eS0)


In [147]:
import pandas as pd

In [148]:
data = pd.read_csv(r"C:\Users\PawPaw\Desktop\Data Science Projects\Weather Data Set\Project 1\file.csv")

In [149]:
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

# How to Analyze Dataframes?

## .head()

It shows the first N rows in the data (By default, N=5)

In [153]:
data.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

# .shape

It shows the total num of rows and num of columns of the dataframe.

In [156]:
data.shape

(8784, 8)

## .index

This attribute provides the index of the dataframe

In [159]:
data.index

RangeIndex(start=0, stop=8784, step=1)

## .columns

This attribute provides the names of each column

In [162]:
data.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

## .dtypes

This shows the data-type of each column

In [165]:
data.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

## unique()

In a column, it shows all the unique values. It can be applied on a single column only, not the whole dataframe.

In [168]:
data['Weather'].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

## .nunique()

It shows the total num of unique values in each column. This can only be applied on a single collumn as well as on the entire data frame

In [171]:
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

## .count

This shows the total num of non-null in each column. It can be applied on a single column as well as the entire dataframe. 

In [174]:
data.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

## .value_counts

In a column, it shows all of the unique values with their count. It can be applied to a single column only. 

In [177]:
data["Weather"].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

## .info()

Provides basic information about the dataframe

In [180]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


# Question 1) Find all the unique "Wind Speed" values in the data

In [182]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [183]:
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

First let's break down what the total number will be. 

In [185]:
data["Wind Speed_km/h"].nunique()

34

Now I want to see the unique values in question

In [187]:
data["Wind Speed_km/h"].unique() # Answer to the question

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2],
      dtype=int64)

From the result above, you'll see that it is indeed 34 unique values in that column. 

# Question 2) Find the number of times when the "Weather is exactly clear"

In [190]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

There are three ways of soving this problem: value counts(), filtering, and groupby().

In [192]:
data.Weather.value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

So I can see that it was clear, 1326 in the data set. Let's try the next way. 

In [194]:
#data.head(2)
data[data.Weather == 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [195]:
#groupby()
data.head(2)
data.groupby("Weather").get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

# Question 3) Find the number of times when the "Wind Speed was exactly 4km/h"

In [197]:
data[data['Wind Speed_km/h'] == 4] #Answer as we can see is 474

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

# Question 4) Find all the Null Values in the data

In [199]:
 data.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

# Question 5) Rename the colum name 'Weather' of the dataframe to "Weather Condition".

In [201]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [202]:
data.rename(columns = {'Weather' : 'Weather Condition'}) # This is the method to change a column's name.


Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

If you wish to make the column name change permenant, then you must add the following to the code seen above:
data.rename(columns = {'Weather' : 'Weather Condition'}, inplace= True) 

# Question 6) What is the mean "Visibilty"?

In [205]:
data.Visibility_km.mean()

27.664446721311478

# Question 7) What is the Standard Deviation of "Pressure" in this data?

In [207]:
data.Press_kPa.std()  #STD is standard deviation

0.8440047459486474

# Question 8) What is the Variance of 'Relative Humidity' in this dataset?

In [209]:
data['Rel Hum_%'].var() # .var provides variance

286.2485501984998

# Question 9) Find all insances when "Snow" was recorded

In [211]:
#Value_counts
data['Weather'].value_counts() #Scroll to the Snow variable and you'll see the exact amount. 

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [212]:
data[data['Weather'] == 'Snow']
#Filtering - At the bottom of this you'll see that there are 390 entries. 

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
55             25.0     100.95    Snow  
84             24.1     101.25    Snow  
86             19.3     100.97    Snow  
87              9.7     100.89    Snow  
88             19.3     100.79    Snow  
...             ...        ...     ...  
8779            9.7     100.13    Snow  
8780            9.7     100.03    Snow  
8781            4.8      99.95    Snow  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow  

[390 rows x 8 columns]

In [213]:
data[data["Weather"].str.contains('Snow')]
#str.contains - This can go deeper and include days of mixed weather such as Snow Showers, etc.

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41      1/2/2012 17:00    -2.1              -9.5         57               22   
44      1/2/2012 20:00    -5.6             -13.4         54               24   
45      1/2/2012 21:00    -5.8             -12.8         58               26   
47      1/2/2012 23:00    -7.4             -14.1         59               17   
48       1/3/2012 0:00    -9.0             -16.0         57               28   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa       Weather  
41             25.0      99.66  Snow Showers  
44             25.0     100.07  Snow Showers  
45             25.0     100.15  Snow Showers  
47             19.3     100.27  Snow Showers  
48             25.0     100.35  Snow Showers  
...             ...        ...           ...  
8779            9.7     100.13          Snow  
8780            9.7     100.03          Snow  
8781            4.8      99.95          Snow  
8782            9.7      99.91          Snow  
8783           11.3      99.89          Snow  

[583 rows x 8 columns]

# Question 10) Find all instances when "When speed is above 24" and "Visability is 25"

In [215]:
data.head(2)  # I pulled this up so I could recall the names of the columns faster

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [216]:
data[(data['Wind Speed_km/h'] > 24) & (data['Visibility_km'] == 25)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa        Weather  
23             25.0      99.31         Cloudy  
24             25.0      99.26   Rain Showers  
25             25.0      99.26         Cloudy  
26             25.0      99.26  Mostly Cloudy  
27             25.0      99.30  Mostly Cloudy  
...             ...        ...            ...  
8705           25.0     101.34   Mainly Clear  
8753           25.0     101.26   Mainly Clear  
8755           25.0     101.47   Mainly Clear  
8759           25.0     101.52  Mostly Cloudy  
8760           25.0     101.51         Cloudy  

[308 rows x 8 columns]

In [217]:
# Two things. First, I focused on getting the two variables I want to measure and then nested them in () in order to add & so I can carry over both 
# measurements. Second, I nested both into their owe data[] to brring them together. 

# Question 11) What is the Mean value of each column against each Weather"

In [219]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [226]:
# Assistant
# Assuming 'data' is a DataFrame and 'Weather' is a categorical column,
# we will select only numeric columns for the mean calculation.

# Group by 'Weather' and calculate the mean for numeric columns only
data.groupby('Weather').mean(numeric_only=True)  # Use numeric_only=True to avoid non-numeric columns

Temp_C  Dew Point Temp_C  \
Weather                                                                
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather                                                               
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        


### Changes made:
- Added `numeric_only=True` to the `mean()` function to ensure that only numeric columns are considered for the mean calculation, preventing the TypeError.

# Question 12) What is the Minimum and Maximum value for each column against each "Weather"

In [233]:
data.groupby("Weather").max()  # To get the minimumn value, it's the same set up but with .min() instead.

Date/Time  Temp_C  \
Weather                                                             
Clear                                       9/9/2012 5:00    32.8   
Cloudy                                     9/9/2012 23:00    30.5   
Drizzle                                    9/30/2012 3:00    18.8   
Drizzle,Fog                                9/30/2012 2:00    19.9   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/19/2012 18:00     1.2   
Drizzle,Snow,Fog                          12/22/2012 3:00     1.1   
Fog                                        9/22/2012 0:00    20.8   
Freezing Drizzle                            2/1/2012 5:00    -2.3   
Freezing Drizzle,Fog                      12/10/2012 5:00    -0.3   
Freezing Drizzle,Haze                      2/1/2012 13:00    -5.0   
Freezing Drizzle,Snow                      3/2/2012 12:00    -3.3   
Freezing Fog                               3/17/2012 6:00    -0.1   
Freezing Rain                               2/1/2012 7:00     0.3   
Freezing Rain,Fog                         12/17/2012 1:00     0.1   
Freezing Rain,Haze                         2/1/2012 15:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      3/13/2012 23:00    14.1   
Mainly Clear                                9/9/2012 9:00    33.0   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             12/27/2012 9:00    -4.9   
Moderate Snow,Blowing Snow               12/27/2012 12:00    -5.4   
Mostly Cloudy                               9/9/2012 2:00    32.4   
Rain                                        9/5/2012 2:00    22.8   
Rain Showers                               9/8/2012 16:00    26.4   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                 12/5/2012 10:00     2.2   
Rain,Fog                                  9/30/2012 23:00    21.7   
Rain,Haze                                  3/13/2012 9:00     5.5   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  4/23/2012 3:00     1.7   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 5:00     1.3   
Snow                                       4/27/2012 9:00     3.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               3/4/2012 21:00     2.9   
Snow Showers,Fog                         12/29/2012 13:00   -10.0   
Snow,Blowing Snow                          2/25/2012 9:00    -1.4   
Snow,Fog                                  3/14/2012 19:00     1.1   
Snow,Haze                                  2/1/2012 21:00    -3.6   
Snow,Ice Pellets                            3/3/2012 4:00     0.8   
Thunderstorms                              7/4/2012 16:00    26.7   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        7/23/2012 18:00    21.3   
Thunderstorms,Rain Showers                  9/8/2012 4:00    25.5   
Thunderstorms,Rain Showers,Fog            7/31/2012 20:00    22.9   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather                                                                
Clear                                                20.4         99   
Cloudy                                               22.6         99   
Drizzle                                              17.7         96   
Drizzle,Fog                                          19.1        100   
Drizzle,Ice Pellets,Fog    

# Question 13) Show all Records where Weather is "Fog"

In [238]:
data[data["Weather"] == 'Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather  
0               8.0     101.24     Fog  
1               8.0     101.24     Fog  
4               4.8     101.23     Fog  
5               6.4     101.27     Fog  
6               6.4     101.29     Fog  
...             ...        ...     ...  
8716            9.7     101.25     Fog  
8717            6.4     101.25     Fog  
8718            9.7     101.25     Fog  
8719            8.0     101.22     Fog  
8722            6.4     101.15     Fog  

[150 rows x 8 columns]

# Question 14) Find all insatnces when "Weather is Clear" or "Visibility" is above 40.

In [242]:
data[(data["Weather"] == "Clear") | (data["Visibility_km"] > 40)]  # to use "Or" you must use |. This is the same set up as Q10.

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
106     1/5/2012 10:00    -6.0             -10.0         73               17   
107     1/5/2012 11:00    -5.6             -10.2         70               22   
108     1/5/2012 12:00    -4.7              -9.6         69               20   
109     1/5/2012 13:00    -4.4              -9.7         66               26   
...                ...     ...               ...        ...              ...   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa        Weather  
67             25.0     101.74          Clear  
106            48.3     100.45   Mainly Clear  
107            48.3     100.41   Mainly Clear  
108            48.3     100.38   Mainly Clear  
109            48.3     100.40   Mainly Clear  
...             ...        ...            ...  
8749           48.3     100.92  Mostly Cloudy  
8750           48.3     100.96   Mainly Clear  
8751           48.3     101.05   Mainly Clear  
8752           48.3     101.15   Mainly Clear  
8756           25.0     101.52          Clear  

[3027 rows x 8 columns]

# Find all the instances when:
## A. Weather is clear and relatie humidity is greater than 50
## or
## B. Visibility is above 40

In [245]:
data[(data["Weather"] == 'Clear') & (data['Rel Hum_%'] > 50) | (data['Visibility_km'] > 40)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
106     1/5/2012 10:00    -6.0             -10.0         73               17   
107     1/5/2012 11:00    -5.6             -10.2         70               22   
108     1/5/2012 12:00    -4.7              -9.6         69               20   
109     1/5/2012 13:00    -4.4              -9.7         66               26   
110     1/5/2012 14:00    -5.1             -10.7         65               22   
...                ...     ...               ...        ...              ...   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa        Weather  
106            48.3     100.45   Mainly Clear  
107            48.3     100.41   Mainly Clear  
108            48.3     100.38   Mainly Clear  
109            48.3     100.40   Mainly Clear  
110            48.3     100.46   Mainly Clear  
...             ...        ...            ...  
8749           48.3     100.92  Mostly Cloudy  
8750           48.3     100.96   Mainly Clear  
8751           48.3     101.05   Mainly Clear  
8752           48.3     101.15   Mainly Clear  
8756           25.0     101.52          Clear  

[2921 rows x 8 columns]